test/integration/grpc_retry_test.cpp

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <gtest/gtest.h>
+
+#include <grpcpp/server_context.h>
+#include <grpcpp/support/status.h>
+#include "framework/test_logger.hpp"
+#include "main/server_runner.hpp"
+#include "network/impl/grpc_channel_builder.hpp"
+#include "qry_responses.pb.h"
+#include "queries.pb.h"
+#include "torii/query_service.hpp"
+
+namespace {
+  constexpr auto kErrorMessage = "this is a test error message";
+};
+
+class MockQueryService : public iroha::torii::QueryService {
+ public:
+  explicit MockQueryService(int max_attempts)
+      : iroha::torii::QueryService(
+          nullptr, nullptr, nullptr, getTestLogger("MockQueryService")),
+        max_attempts_(max_attempts) {}
+
+  grpc::Status Find(grpc::ServerContext *context,
+                    const iroha::protocol::Query *request,
+                    iroha::protocol::QueryResponse *response) override {
+    if (attempts_ < max_attempts_) {
+      attempts_++;
+      return grpc::Status(grpc::StatusCode::UNAVAILABLE, kErrorMessage);
+    }
+    return grpc::Status::OK;
+  }
+
+ private:
+  int attempts_ = 0;
+  int max_attempts_;
+};
+
+namespace {
+  constexpr int kAttemptsForSuccess = 4;  // 5 max attempts at reaching
+  constexpr int kAttemptsForFailure = 5;  // 5 max attempts at reaching
+  constexpr auto kListenIP = "127.0.0.1";
+  constexpr auto kListenPort = 31337;
```

Comment:
It should be possible to use `0` so that it will assign any port that is not used.